src/main/java/io/supertokens/session/jwt/JWT.java

```diff
@@ -77,15 +87,30 @@ public static JWTInfo verifyJWTAndGetPayload(String jwt, String publicSigningKey
         } catch (InvalidKeySpecException | SignatureException e) {
             throw new JWTException("JWT verification failed");
         }
-        return new JWTInfo(new JsonParser().parse(Utils.convertFromBase64(splittedInput[1])),
-                splittedInput[0].equals(JWT.HEADERv1) ? AccessToken.VERSION.V1 : AccessToken.VERSION.V2);
+        return new JWTInfo(new JsonParser().parse(Utils.convertFromBase64(splittedInput[1])), version);
+    }
+
+    private static AccessToken.VERSION parseVersionHeader(String header) {
+        if (header.equals(JWT.HEADERv1)) {
+            return AccessToken.VERSION.V1;
+        } else if (header.equals(JWT.HEADERv2)) {
+            return AccessToken.VERSION.V2;
+        } else if (header.equals(JWT.HEADERv3)) {
+            return AccessToken.VERSION.V3;
+        }
+
+        return null;
     }
 
     public static JWTInfo getPayloadWithoutVerifying(String jwt) {
         initHeader();
         String[] splittedInput = jwt.split("\\.");
-        return new JWTInfo(new JsonParser().parse(Utils.convertFromBase64(splittedInput[1])),
-                splittedInput[0].equals(JWT.HEADERv1) ? AccessToken.VERSION.V1 : AccessToken.VERSION.V2);
+        AccessToken.VERSION version = parseVersionHeader(splittedInput[0]);
+        if (version == null) {
+            version = AccessToken.VERSION.V3;
+        }
```

Comment:
Why would this ever be null?

Reply:
When we are trying to validate an access token that has an invalid header.